solr/core/src/test/org/apache/solr/update/TestNestedUpdateProcessor.java

```diff
@@ -0,0 +1,189 @@
+      "                          },\n" +
+      "                },\n" +
+      "                {\n" +
+      "                    \"foo_s\": \"Bar\"\n" +
+      "                }\n" +
+      "            ]\n" +
+      "        }\n" +
+      "    }\n" +
+      "}";
+
+  private static final String errDoc = "{\n" +
+      "    \"add\": {\n" +
+      "        \"doc\": {\n" +
+      "            \"id\": \"1\",\n" +
+      "            \"children" + PATH_SEP_CHAR + "a\": [\n" +
+      "                {\n" +
+      "                    \"id\": \"2\",\n" +
+      "                    \"foo_s\": \"Yaz\"\n" +
+      "                    \"grandChild\": \n" +
+      "                          {\n" +
+      "                             \"id\": \""+ grandChildId + "\",\n" +
+      "                             \"foo_s\": \"Jazz\"\n" +
+      "                          },\n" +
+      "                },\n" +
+      "                {\n" +
+      "                    \"id\": \"3\",\n" +
+      "                    \"foo_s\": \"Bar\"\n" +
+      "                }\n" +
+      "            ]\n" +
+      "        }\n" +
+      "    }\n" +
+      "}";
+
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  @BeforeClass
+  public static void beforeClass() throws Exception {
+    initCore("solrconfig-update-processor-chains.xml", "schema15.xml");
+  }
+
+  @Before
+  public void before() throws Exception {
+    assertU(delQ("*:*"));
+    assertU(commit());
+  }
+
+  @Test
+  public void testDeeplyNestedURPGrandChild() throws Exception {
+    indexSampleData(jDoc);
+
+    assertJQ(req("q", IndexSchema.PATH_FIELD_NAME + ":*" + PATH_SEP_CHAR + "grandChild",
+        "fl","*",
+        "sort","id desc",
+        "wt","json"),
+        "/response/docs/[0]/id=='" + grandChildId + "'");
+  }
+
+  @Test
+  public void testDeeplyNestedURPChildren() throws Exception {
+    final String[] childrenTests = {"/response/docs/[0]/id=='" + childrenIds[0] + "'", "/response/docs/[1]/id=='" + childrenIds[1] + "'"};
+    indexSampleData(jDoc);
+
+    assertJQ(req("q", IndexSchema.PATH_FIELD_NAME + ":children",
+        "fl","*",
+        "sort","id asc",
+        "wt","json"),
+        childrenTests);
+  }
+
+  @Test
+  public void testDeeplyNestedURPChildrenWoId() throws Exception {
+    final String[] childrenTests = {"/response/docs/[0]/id=='1" + PATH_SEP_CHAR + "children" + PATH_SEP_CHAR + "2'"};
+    indexSampleData(noIdChildren);
+
+    assertJQ(req("q", "foo_s:Bar",
+        "fl","*",
+        "sort","id asc",
+        "wt","json"),
+        childrenTests);
+  }
+
+  @Test
+  public void testDeeplyNestedURPFieldNameException() throws Exception {
+    final String errMsg = "contains: '" + PATH_SEP_CHAR + "' , which is reserved for the nested URP";
+    thrown.expect(SolrException.class);
+    indexSampleData(errDoc);
+    thrown.expectMessage(errMsg);
+  }
+
+  private void indexSampleData(String cmd) throws Exception {
```

Comment:
I'm skeptical we need to write this much code for something routine.  See `updateJ()`; can you just call that?

Reply:
Sorry, kind of missed that method.
My bad.